Automated Smoketest: :x:
OS: centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: 826707715e51a863e9aafc6dad2129fbc57457e0
Number of Error(s): 5
/home/centos/smoketest/PR-14127/HPCC-Platform/rtl/eclrtl/rtlfield.cpp:3231:1: error: control reaches end of non-void function [-Werror=return-type]
cc1plus: all warnings being treated as errors
make[2]: *** [rtl/eclrtl/CMakeFiles/eclrtl.dir/build.make:245: rtl/eclrtl/CMakeFiles/eclrtl.dir/rtlfield.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:16674: rtl/eclrtl/CMakeFiles/eclrtl.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2020-09-02_12-49-47
time logs started
Prep time   :     8 sec (00:00:08)
Build time  :   196 sec (00:03:16)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   204 sec (00:03:24)
time logs end

[Build error:](https://gist.github.com/a99370f2040178b3b4bc409729631155)
Build log [Build log](https://gist.github.com/HPCCSmoketest/a99370f2040178b3b4bc409729631155#file-result-20-09-02-12-45-45-log)

--- a/rtl/eclrtl/rtlfield.cpp
+++ b/rtl/eclrtl/rtlfield.cpp
@@ -3192,6 +3192,45 @@ extern bool ECLRTL_API hasTrailingFileposition(const RtlTypeInfo * type)
     return false;
 }
 
+//Does the field list contain a keyed int?  We are not interested in fields inside child datasets.
+//And keyed fields cannot exist within ifblocks.
+static bool containsKeyedSignedInt(const RtlFieldInfo * const * fields)
+{
+    if (!*fields)
+        return false;
+    while (*fields)
+    {
+        const RtlTypeInfo * type = (*fields)->type;
+        if (type->getType() == type_keyedint)
+        {
+            const RtlTypeInfo * baseType = type->queryChildType();
+            if (baseType->isSigned() && (baseType->getType() == type_int))
+                return true;
+        }
+        else if (type->getType() == type_record)
+        {
+            const RtlRecordTypeInfo * record = static_cast<const RtlRecordTypeInfo *>(type);
+            if (containsKeyedSignedInt(record->fields))
+                return true;
+        }
+        fields++;
+    }
+    return false;
+}
+
+extern bool ECLRTL_API containsKeyedSignedInt(const RtlTypeInfo * type)
+{
+    switch (type->getType())
+    {
+    case type_record:
+    {
+        const RtlRecordTypeInfo * record = static_cast<const RtlRecordTypeInfo *>(type);
+        return containsKeyedSignedInt(record->fields);
+    }
+    }
+    return false;
+}
+
 //-------------------------------------------------------------------------------------------------------------------
 
 size32_t RtlRecordTypeInfo::getMinSize() const

--- a/rtl/eclrtl/rtlfield.hpp
+++ b/rtl/eclrtl/rtlfield.hpp
@@ -752,6 +752,8 @@ extern int ECLRTL_API compareFields(const RtlFieldInfo * const * cur, const byte
 extern unsigned ECLRTL_API hashFields(const RtlFieldInfo * const * cur, const byte *self, unsigned inhash, bool excludePayload = false);
 extern bool ECLRTL_API hasTrailingFileposition(const RtlFieldInfo * const * fields);
 extern bool ECLRTL_API hasTrailingFileposition(const RtlTypeInfo * type);
+extern bool ECLRTL_API containsKeyedSignedInt(const RtlTypeInfo * type);
+
 extern size32_t translateScalar(ARowBuilder &builder, size32_t offset, const RtlFieldInfo *field, const RtlTypeInfo &destType, const RtlTypeInfo &sourceType, const byte *source);
 
 #endif

--- a/rtl/eclrtl/rtlnktest.cpp
+++ b/rtl/eclrtl/rtlnktest.cpp
@@ -311,6 +311,10 @@ protected:
     {
         RtlStringTypeInfo str1(type_string, 1);
         RtlIntTypeInfo int2(type_int, 2);
+        RtlIntTypeInfo int8(type_int, 8);
+        RtlSwapIntTypeInfo swapint2(type_int, 2);
+        RtlSwapIntTypeInfo swapint8(type_int, 8);
+        RtlKeyedIntTypeInfo keyedint8(type_keyedint, 8, &swapint8);
         RtlStringTypeInfo strx(type_string|RFTMunknownsize, 0);
         RtlUtf8TypeInfo utf8(type_utf8|RFTMunknownsize, 0, nullptr);
         testSerialize(int2, "[123]");
@@ -322,6 +326,23 @@ protected:
         testSerialize(int2, "(,)");
         testSerialize(int2, "(123,234),(456,567)");
         testSerialize(int2, "(456,567),(123,234)", "(123,234),(456,567)");
+
+        testSerialize(swapint2, "[123]");
+        testSerialize(swapint2, "(123,234]");
+        testSerialize(swapint2, "[123,234)");
+        testSerialize(swapint2, "(123,234)");
+        testSerialize(swapint2, "(,234)");
+        testSerialize(swapint2, "(128,)");
+        testSerialize(swapint2, "(,)");
+        testSerialize(swapint2, "(123,234),(456,567)");
+        testSerialize(swapint2, "(456,567),(123,234)", "(123,234),(456,567)");
+
+        testSerialize(int8, "[9223372036854775807]");
+        testSerialize(int8, "[-9223372036854775808]");
+        testSerialize(swapint8, "[9223372036854775807]");
+        testSerialize(swapint8, "[-9223372036854775808]");
+        testSerialize(keyedint8, "[10]");
+        testSerialize(keyedint8, "[-10]");
 
         testSerialize(str1, "['A']");
         testSerialize(str1, "[',']");

--- a/thorlcr/activities/indexread/thindexreadslave.cpp
+++ b/thorlcr/activities/indexread/thindexreadslave.cpp
@@ -173,7 +173,8 @@ public:
                     Owned<ITranslator> translator = getTranslators(part);
                     IOutputMetaData *actualFormat = translator ? &translator->queryActualFormat() : expectedFormat;
                     bool tryRemoteStream = actualFormat->queryTypeInfo()->canInterpret() && actualFormat->queryTypeInfo()->canSerialize() &&
-                                           projectedFormat->queryTypeInfo()->canInterpret() && projectedFormat->queryTypeInfo()->canSerialize();
+                                           projectedFormat->queryTypeInfo()->canInterpret() && projectedFormat->queryTypeInfo()->canSerialize() &&
+                                           !containsKeyedSignedInt(actualFormat->queryTypeInfo());
 
                     /* If part can potentially be remotely streamed, 1st check if any part is local,
                      * then try to remote stream, and otherwise failover to legacy remote access

